cmd/antrea-controller/controller.go

```diff
@@ -114,33 +123,89 @@ func run(o *Options) error {
 	return nil
 }
 
-func createAPIServerConfig(kubeconfig string,
+func createAPIServerConfig(client clientset.Interface,
+	aggregatorClient aggregatorclientset.Interface,
+	selfSignedCert bool,
+	kubeconfig string,
 	bindPort int,
 	addressGroupStore storage.Interface,
 	appliedToGroupStore storage.Interface,
 	networkPolicyStore storage.Interface,
 	controllerQuerier querier.ControllerQuerier,
 	enableMetrics bool) (*apiserver.Config, error) {
-	// TODO:
-	// 1. Support user-provided certificate.
 	secureServing := genericoptions.NewSecureServingOptions().WithLoopback()
 	authentication := genericoptions.NewDelegatingAuthenticationOptions()
 	authorization := genericoptions.NewDelegatingAuthorizationOptions()
 
-	// Set the PairName but leave certificate directory blank to generate in-memory by default
-	secureServing.ServerCert.CertDirectory = ""
-	secureServing.ServerCert.PairName = "antrea-apiserver"
+	var certPEMBlock []byte
+	var err error
+	if selfSignedCert {
+		// Set the PairName but leave certificate directory blank to generate in-memory by default
+		secureServing.ServerCert.CertDirectory = ""
+		secureServing.ServerCert.PairName = "antrea-apiserver"
+
+		if err := secureServing.MaybeDefaultWithSelfSignedCerts("localhost", []string{"antrea.kube-system.svc", "antrea.kube-system.svc.local.cluster"}, []net.IP{net.ParseIP("127.0.0.1")}); err != nil {
+			return nil, fmt.Errorf("error creating self-signed certificates: %v", err)
+		}
+		certPEMBlock = pem.EncodeToMemory(&pem.Block{Type: "CERTIFICATE", Bytes: secureServing.ServerCert.GeneratedCert.Certificate[0]})
+	} else {
+		// TODO: Watch these files and restart apiserver if changed.
+		secureServing.ServerCert.CertKey.CertFile = "/var/run/antrea/secrets/tls.crt"
+		secureServing.ServerCert.CertKey.KeyFile = "/var/run/antrea/secrets/tls.key"
+		certPEMBlock, err = ioutil.ReadFile("/var/run/antrea/secrets/ca.crt")
+		if err != nil {
+			return nil, fmt.Errorf("error reading CA certificate: %v", err)
+		}
+	}
+
+	exists := true
+	caConfigMap, err := client.CoreV1().ConfigMaps("kube-system").Get("antrea-ca", metav1.GetOptions{})
```

Comment:
would it be a good idea to create an empty config map as part of the yaml? We could simplify the code here and simplify RBAC for the controller (no longer need the ability to create config maps, just need to be able to update the `antrea-ca` config map).

Reply:
I like the more restricted RBAC, and I think we can block here with a log message. But no strong preference. Maybe @jianjuns has an opinion on this?

Reply:
No strong opinion either, but do you feel strange controller/agent will fail if the configmap is not created in advance. This is the only case we depend on resource created via YAML?

Reply:
If we rely on deployment yaml to create the configmap, maybe we don't need to worry the case that configmap is not created in advance, because many functionalities have depended on the correct yaml being applied: RBAC, service discovery (we assume antrea service is created in advance), environment variable.
Now I prefer creating the configmap via the yaml too, and just reading the configmap from api, as even lib apiserver is reading CA from configmap api:
https://github.com/kubernetes/kubernetes/blob/release-1.15/staging/src/k8s.io/apiserver/pkg/server/options/authentication.go#L195.
and 1.17 provides a class to get latest CA bundle from configmap near-realtime:
https://github.com/kubernetes/kubernetes/blob/release-1.17/staging/src/k8s.io/apiserver/pkg/server/options/authentication.go#L269.
I feel our usage is quite similar to this and I can reuse the class to retrieve CA from our configmap after bumping up K8s API version

Reply:
Ok. Sounds good to me.